src/expr.rs

```diff
@@ -53,6 +54,54 @@ pub(crate) fn format_expr(
     expr_type: ExprType,
     context: &RewriteContext<'_>,
     shape: Shape,
+) -> Option<String> {
+    // when max_width is tight, we should check all possible formattings, in order to find
+    // if we can fit expression in the limit. Doing it recursively takes exponential time
+    // relative to input size, and people hit it with rustfmt takes minutes in #4476 #4867 #5128
+    // By memoization of format_expr function, we format each pair of expression and shape
+    // only once, so worst case execution time becomes O(n*max_width^3).
+    if context.inside_macro() || context.is_macro_def {
+        // span ids are not unique in macros, so we don't memoize result of them.
+        return format_expr_inner(expr, expr_type, context, shape);
+    }
+    let clean;
+    let query_id = QueryId {
+        shape,
+        span: expr.span,
+    };
+    if let Some(map) = context.memoize.take() {
+        if let Some(r) = map.get(&query_id) {
+            let r = r.clone();
+            context.memoize.set(Some(map)); // restore map in the memoize cell for other users
```

Comment:
This is the piece I'm struggling a bit to conjure up mentally. If we've got the final formatting for a specific expression, in what context would we need to get that particular entry again given the hierarchical structure/who would other users be?

Reply:
In the recursive calls it will be called again if formating fails due exceeding max_width limit:
```
loooooooooooong(looooooooooong(this_one_formats(twice)))
```
calculate formatting of `this_one_formats(twice)` but it fails later, and we try this:
```
loooooooooooong(
    looooooooooong(
         this_one_formats(twice)
   )
)
```
this formats `this_one_formats(twice)` identical to above, but calculate it independently.

So other users here are other recursive calls to the `format_expr` functions.

Reply:
discussion shifted outside this comment so resolving